src/spesh/optimize.c

```diff
@@ -1065,60 +951,182 @@ static void optimize_istrue_isfalse(MVMThreadContext *tc, MVMSpeshGraph *g, MVMS
+                ins->info = MVM_op_get_op(MVM_OP_sp_boolify_iter_hash);
+            } else {
+                ins->info = MVM_op_get_op(MVM_OP_sp_boolify_iter);
+            }
+            break;
+        case MVM_BOOL_MODE_UNBOX_INT:
+            if (!guaranteed_concrete)
+                return;
+                /* We can just unbox the int and pretend it's a bool. */
+            ins->info = MVM_op_get_op(MVM_OP_unbox_i);
+            /* And then we might be able to optimize this even further. */
+            optimize_unbox(tc, g, bb, ins);
+            break;
+        case MVM_BOOL_MODE_BIGINT:
+            if (!guaranteed_concrete)
                 return;
+            ins->info = MVM_op_get_op(MVM_OP_bool_I);
+            break;
+        case MVM_BOOL_MODE_HAS_ELEMS:
+            if (!guaranteed_concrete)
+                return;
+            ins->info = MVM_op_get_op(MVM_OP_elems);
+            optimize_repr_op(tc, g, bb, ins, 1);
+            break;
+        case MVM_BOOL_MODE_NOT_TYPE_OBJECT:
+            ins->info = MVM_op_get_op(MVM_OP_isconcrete);
+            /* And now defer another bit of optimization */
+//            optimize_isconcrete(tc, g, ins);
+            break;
+            /* We need to change the register type for our result for this,
+             * means we need to insert a temporarary and a coerce:
+        case MVM_BOOL_MODE_UNBOX_NUM:
+             op_info = MVM_op_get_op(MVM_OP_unbox_i);
+             break;
+            */
+        default:
+            return;
         }
-        /* Now we can take care of the negation. */
+//        fprintf(stderr, "Known type optimized to %s\n", ins->info->name);
+        /* Now we can take care of the negation. - NB I'm not entirely sure why
+         * this would need it's own register though! */
         if (negated_op) {
             /* Insert a not_i instruction that negates temp. This not_i is
              * subject to further optimization in the case that temp has a
              * known value set on it. */
+            MVMSpeshOperand orig       = ins->operands[0];
+            MVMSpeshOperand temp       = MVM_spesh_manipulate_get_temp_reg(tc, g, MVM_reg_int64);
             MVMSpeshIns     *new_ins   = MVM_spesh_alloc(tc, g, sizeof( MVMSpeshIns ));
             MVMSpeshOperand *operands  = MVM_spesh_alloc(tc, g, sizeof( MVMSpeshOperand ) * 2);
+            MVMSpeshFacts  *temp_facts = MVM_spesh_get_facts(tc,g,temp);
+/*          fprintf(stderr, "inserting not_i to negate istrue\n"); */
             new_ins->info = MVM_op_get_op(MVM_OP_not_i);
             new_ins->operands = operands;
             operands[0] = orig;
             operands[1] = temp;
+            ins->operands[0] = temp;
             MVM_spesh_manipulate_insert_ins(tc, bb, ins, new_ins);
-            MVM_spesh_get_facts(tc, g, temp)->usages++;
             MVM_spesh_manipulate_release_temp_reg(tc, g, temp);
+
+            /* Set facts on temporary - NB this is *very* confusing */
```

Comment:
Why?

Reply:
It did something different before, that was also wrong :smile: